### ISSUE TITLE
Support Jupyter Notebook Interactive Computing paradigm

### ISSUE THREAD

**Issue by FuriouslyCurious** [Microsoft/vscode#34739]
Around 500,000 VSCode users have been relying on Don Jayamanne has "vscodeJupyter" extension to connect to Jupyter Notebooks and execute code in an interactive environment.  (https://github.com/DonJayamanne/vscodeJupyter)

With rapid continuing growth of Python, Data Science, and Deep Learning communities, Jupyter Notebooks and interactive computing need to be promoted to a first class citizen of VS Code, rather than being an extension developed by a volunteer.  

Please refer to Atom + Hydrogen integration, and upcoming JupyerLab projects for ideas on how to integrate Notebooks inside Visual Studio Code.

Thank you!

**Comment by MikhailArkhipov** [comment]
Related https://github.com/Microsoft/vscode/issues/39492
@DonJayamanne now works for Microsoft BTW :-)

**Comment by FuriouslyCurious** [comment]
That's awesome @MikhailArkhipov and congratulations @DonJayamanne 

We hope to see more amazing REPL, Interactive Computing, and Pythonic things from vsCode now.

**Comment by domschl** [comment]
See: #3220

**Comment by KhurrumMahmood** [comment]
An analog of Atom's Hydrogen or Swift's Playground would be very promising (or just a smoother implementation of the current extension).

MS should ask the current users of the extension for feedback to see what improvements people would be interested in - apart from the obvious.

A roadmap would be good too, so people can give feedback on it.

**Comment by yoDon** [comment]
@nojvek also has a repo with VS Code support for converting between .py and .ipynb on the fly

https://github.com/nojvek/vscode-ipynb-py-converter

**Comment by jflam** [comment]
I had some students at ICL build a prototype of this idea - here's their plugin: https://marketplace.visualstudio.com/items?itemName=neuron.neuron-IPE 

Feedback welcome!

**Comment by jamescasbon** [comment]
I agree that first class notebook support would be massive, and I looked into adding it with existing extension points.  However, I don't think its possible.

Motivation: many data scientists use ipython notebook as remote execution environment.  However, using the notebook separate to existing ide introduces friction when trying to do the most obvious tasks (eg moving a function prototyped in a notebook into library code).  It also means using a separate editor/debugger setup.  Currently, jupyter is adding more and more ide features - but first class support for notebooks from an ide would be a lot better. 

Suggestion, we need an vs code extension that can: 
 - render the notebook format [1] https://ipython.org/ipython-doc/3/notebook/nbformat.html showing alternating source input and html output cells.  
 - execute code on a notebook and update the view [2]

There is support for both of these individually via plugins.  With the following drawbacks:

 [1] is done via virtual documents or webviews,  I believe neither is capable of presenting an editable view of the notebook.  I looked into the filesystemprovider extension point.  Using this it would be possible to transform the raw json notebook into a reasonable looking editable file with interleaved input and output.  It could then use the code lens to present actions for each cell.  However, this still doesn't quite do enough:  the output cells can contain html (eg inline images) and I don't see a way of rendering inline html in a source document.

[2] executing the code and putting in a virtual document is already working well in other plugins.  Therefore, this part should be simple enough given support for [1].


See also #41775

**Comment by gramster** [comment]
Note that there is an interactive experience in the Python extension now which includes import/export to .ipynb.

**Comment by xgdgsc** [comment]
Importing / exporting feels weired, I don' t want another file.

**Comment by stonebig** [comment]
I did some experience, and on my win10 pc, Jupyter seems about 2 times slower at first try when used from vscode 
 ... is it possible/normal  ?

from vscode
![image](https://user-images.githubusercontent.com/4312421/63769395-ffb12e80-c8d2-11e9-8d0a-94f7dd492461.png)

from a jupyter notebokk, the same code
![image](https://user-images.githubusercontent.com/4312421/63778511-3ee87b00-c8e5-11e9-9596-e1f22b83c22d.png)

.

**Comment by haberdashPI** [comment]
Not sure if this is the right place to comment, if this should be a new issue:

Python is not the only language Jupyter was intended for: Julia and R are both first class citizens of the framework: (ju)lia-(pyt)hon-(r). There are vscode extensions for both R and Julia. Ultimately it would be *awesome* if the support being added for Jupyter could have hooks so these extensions can also leverage notebooks.

**Comment by ejazr83** [comment]
I would like to expan on the above feature request to allow use of remote development feature of VScode when using remote Jupyter notebooks

Currently, to be able to execute remote Jupyter notebooks in VSCode, you need to have the Python runtime installed on your local machine. If VSCode could use the remote dev features, then other runtimes like R, Julia etc could run in the VSCode context as well.

To get the same benefits of remote VS Code development such as not requiring to install Python runtime, the Python Jupter remote execution should be able to run similar to a remote VS Code development environment.

Currently this works when you SSH to a box but it would be helpful if this can also be initiated when a remote Jupyter notebook is connected to so that the cells are run remotely rather than in local runtime

This will effectively make VS code use the remote runtime for the Jupyter notebooks rather than the local runtime removing the need to have Python installed locally for this use case.

**Comment by alekseiAlefirov** [comment]
I would like to agree with that Notebooks should not be directly associated with programming in Python in VS Code. Considering that [the community contributes a lot in creating different language kernels for Jupyter](https://github.com/jupyter/jupyter/wiki/Jupyter-kernels), it seems natural to have this awesome Jupyter front-end, that now resides in the Python extension, extracted. It would be really great to have a generic mechanism to provide possibility **to write Notebooks in any programming language in VS Code**.

**Comment by gramster** [comment]
See https://github.com/microsoft/vscode/issues/84293
